### PR TITLE
add YouCanBootAnythingAsLongAsItsAlpine image source

### PR DESCRIPTION
**Pull request by jmpesp**
For development purposes, allow people to select the alpine.iso blob that ships with the propolis-server zone image as an image source.

### CODE REVIEW

File: nexus/src/external_api/params.rs [hunk 1]
```diff
@@ -409,6 +409,7 @@ pub struct NetworkInterfaceIdentifier {
 pub enum ImageSource {
     Url(String),
     Snapshot(Uuid),
+    File(String),
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/external_api/params.rs]: the ability to specify arbitrary paths was ripped out

File: nexus/src/nexus.rs [hunk 2]
```diff
@@ -1239,6 +1239,107 @@ impl Nexus {
...
+                let digest = hex::encode({
+                    use ring::digest::{Context, SHA256};
+                    use std::fs::File;
+                    use std::io::{BufReader, Read};
+
+                    let mut context = Context::new(&SHA256);
+                    let mut buffer = [0; 1024];
+
+                    let input =
+                        File::open(&path).map_err(|e| Error::InvalidValue {
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/nexus.rs]: the ability to specify arbitrary paths was ripped out

**Comment by jmpesp** [comment]
Unfortunately this doesn't seem to boot through through the default NVMe device emulation. This can be reproduced with:
    
        wget https://oxide-omicron-build.s3.amazonaws.com/OVMF_CODE.fd
        wget https://oxide-omicron-build.s3.amazonaws.com/alpine.iso
        qemu-system-x86_64 \
            -bios OVMF_CODE.fd \
            -machine pc-i440fx-focal \
            -m 1g \
            -drive file=alpine.iso,if=none,format=raw,id=nvme1 \
            -device nvme,drive=nvme1,serial=nvme-1,addr=0x4

**Comment by jmpesp** [comment]
Note the original propolis config used the "pci-virtio-block" device.

**Comment by jmpesp** [comment]
> Unfortunately this doesn't seem to boot through through the default NVMe device emulation.

I should be more clear - it does boot to a UEFI shell. If you enter `FS0:\efi\boot\bootx64.efi`, it does take you to a grub prompt, and fails to boot further. On the serial console:

```
Mounting boot media failed. 
initramfs emergency recovery shell launched. Type 'exit' to continue boot
sh: can't access tty; job control turned off
/ #
```

I suspect this is due to a missing NVMe module.

**Comment by jmpesp** [comment]
The latest [alpine-virt-3.16.0-x86_64.iso](https://dl-cdn.alpinelinux.org/alpine/v3.16/releases/x86_64/alpine-virt-3.16.0-x86_64.iso) does boot with the qemu invocation above (specifically, through NVMe).

**Comment by jmpesp** [comment]
I've updated our alpine.iso to alpine-virt-3.16.0-x86_64.iso, and now the above qemu invocation works, and booting an instance using a disk based on

```
~/cli/cli.sh api /images --method POST --input - <<EOF
{
  "name": "alpine",
  "description": "boot from propolis zone blob!",
  "block_size": 512,
  "source": {
    "type": "you_can_boot_anything_as_long_as_its_alpine"
  }
}
EOF
```

works too.

**Review by smklein — APPROVED** [review]

**Pull request merged**